spec/formatting.md

```diff
@@ -690,20 +707,20 @@ One such strategy MUST behave as follows:
    1. Let `fmt` be the formatted string representation of the resolved value of `exp`.
    1. Let `dir` be the directionality of `fmt`,
       one of « `'LTR'`, `'RTL'`, `'unknown'` », with the same meanings as for `msgdir`.
-   1. If `dir` is `'unknown'`:
+   1. If `dir` is `'RTL'`:
       1. In the formatted output,
-         prefix `fmt` with U+2068 FIRST STRONG ISOLATE
+         prefix `fmt` with U+2067 RIGHT-TO-LEFT ISOLATE
          and postfix it with U+2069 POP DIRECTIONAL ISOLATE.
-   1. Else, if `dir` is `'LTR'` and `msgdir` is not `'LTR'`:
```

Comment:
So just to clarify, you would like us to change the default bidi isolation strategy to _always_ add isolates around placeholders, even if they match the message's base direction? So for example when formatting the message `Hello {42 :number}!` in English, the expected output would be:
```
Hello \u{2066}42\u{2069}!
```

Could you give an example for what sort of situation and message this change is necessary for?

Reply:
The challenge here is to get isolation when it is needed without introspecting the characters in the pattern and placeholder string. Spillover can occur for a number of reasons, but it requires a string that has mixed direction and it most frequently involves cases where the placeholder and adjacent text have the same direction (so that there is no directional boundary) or where neutral or weakly directional characters are involved.

Numbers (especially currency values) and dates often have weakly directional or opposite directional characters in them and they are common in placeholders. An example of this is [Example 13](https://w3c.github.io/i18n-discuss/explainers/bidi-html-demo.html?item=13), where RTL placeholders for two currency values have a neutral between them:

![image](https://github.com/unicode-org/message-format-wg/assets/69082/40a7e346-41b8-4547-b4ed-e4e87d1c3734)


We could require RTL patterns to do isolation all the time (because, excepting Adlam, numbers are weakly LTR in most such languages) and exempt LTR patterns for LTR placeholders. Or we could allow/require implementations to introspect. If there is no mixing of directional runs, there can't be spillover. Let me make some additional edits...

Reply:
How would that look like in an MF2 message pattern? I'm struggling to figure out which parts of the example are the nominally RTL parts that need isolation.

Reply:
The Arabic means "The price is {$price :currency} + {$shipping :currency} shipping".

The first value is "1234.56 USD". The second is "12.99 AED" (which currency is used doesn't matter). Both are formatted for the ar-AE locale in the example and thus have a base direction of RTL (the "USD" should appear to the left of the number because it is in a trailing position). The appearance of LTR characters in a formatted RTL string is not unusual.

The problem here is that `+` is a neutral, so you end up with the weak `12.99` being attached to the end of the LTR run. If a `dir=auto` or `dir=rtl` were added to the `<span>` elements, the message presents correctly. FWIW, message patterns need isolation to be legible in RTL:

![image](https://github.com/unicode-org/message-format-wg/assets/69082/ea32e616-b372-4e68-9436-3becc7cfac17)

Reply:
Check out the change to the algorithm. Basically I allow implementations to omit isolates when the message and locale are both LTR (which is the most common case).